docs/design/sql/20-matching-sql-query-to-jet-job.md

```diff
@@ -0,0 +1,237 @@
+where upper 128 bit represents the query coordinator, and lower 128 bit represents the actual query id, which is
+generated by `UuidUtil.newUnsecureUUID()`. On the client, it should be stored locally in set or map, and used by
+metrics consumer.
+
+The basic idea of this approach is simple : we may take lower 64 bits of `QueryId` and use them as **Jet job**
+identifier during job creation in `PlanExecutor`:
+
+java
+class PlanExecutor {
+// ...
+SqlResult execute(SelectPlan plan, QueryId queryId, List<Object> arguments, long timeout, SqlSecurityContext ssc) {
+// ...
+long jobId = queryId.getLocalIdLow();
+// ...
+Job job = jet.newLightJob(jobId, plan.getDag(), jobConfig, ssc.subject());
+}
+// ...
+}
+
+One significant drawback of this approach, which cannot be easily addressed, is that Jet depends on a cluster-wide
+flake ID generator to produce job IDs, thereby ensuring their uniqueness across the entire cluster. However, if
+the client utilizes `UuidUtil.newUnsecureUUID()`, it cannot guarantee the same level of uniqueness. To prevent
+ID collisions, our solution involves storing all job IDs in a cluster-wide, consistent data store. Regrettably,
+the Hazelcast CP subsystem is not available for the SQL engine and Jet.
+
+In this scenario, we must consider the possibility that the client may be under the control of potentially malicious
+parties.
+
+### Approach 2 : return `jobId` to the client as a part of `SqlResult`
+
+This approach differs from the previous one in that it doesn't use `QueryId` at all. Instead, we are making an
+additional field in `SqlResult` (or `SqlClientResult`) class, which will contain `jobId` of the corresponding Jet job.
+
+This seemingly minor alteration, when examined closely, carries significant implications:
+we must now match Jet jobs with SQL queries during query execution, ideally even before the Jet job start its execution.
+The challenge with this approach is that presently (as of HZ 5.3), for small batch queries, as well as for all DML
+queries, the `SqlResult` is not available until after the job has finished. To address this predicament, we need
+to rework our query lifecycle to operate asynchronously: we'll return an empty `SqlResult` (or empty analogue
+of `SqlResult`) to the client and prompt the client to `fetch` the results. Worth noticing, it also solves a problem
+of query thread blocking during DML query execution.
+
+The primary difficulty with this method lies in the fact that, when executing DML queries, we cannot ensure
+the existence of the job entity after receiving the job ID as part of a SQL result. This means that if the query fails,
+the end user won't be able to determine its status – whether it was successful or not. This is not acceptable behaviour,
+and this limitation encouraged us to explore an alternative solution:
+
+### Approach 3 : return `jobId` to the client as a part of `SqlExecuteResponce`
```

Comment:
`SqlExecuteResponce` is `SqlResult`. This is option2 + ability to get final result of DML by queryId

Reply:
you do not mention queryId explicitly in the description not the use of `SqlFetchMessageTask`